File: src/util/npm-install.ts
import * as os from "node:os";
import * as path from "node:path";
import * as fs from "node:fs";
import { run } from "./exec.js";

const REGISTRY = "https://registry.npmjs.org/";

async function resolveFromRegistry(
  pkg: string,
  spec: string,
): Promise<{ version: string; tarball: string } | null> {
  try {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), 5000);
    const res = await fetch(`${REGISTRY}${encodeURIComponent(pkg)}`, { signal: ctrl.signal });
    clearTimeout(timer);
    if (!res.ok) return null;
    const doc = (await res.json()) as {
      "dist-tags"?: Record<string, string>;
      versions?: Record<string, { version: string; dist?: { tarball?: string } }>;
    };
    const version = doc["dist-tags"]?.[spec] ?? (doc.versions?.[spec] ? spec : null);
    if (!version) return null;
    const tarball = doc.versions?.[version]?.dist?.tarball;
    if (!tarball) return null;
    return { version, tarball };
  } catch {
    return null;
  }
}

// Universal, cache-skew-resistant global install for an npm package.
export async function npmGlobalInstall(pkg: string, spec = "latest"): Promise<string | null> {
  const resolved = await resolveFromRegistry(pkg, spec);
  const target = resolved?.version ?? null;
  const atSpec = `${pkg}@${spec}`;

  const attempts: string[][] = [
    ["install", "-g", atSpec],
    ["install", "-g", atSpec, "--prefer-online"],
    ["install", "-g", atSpec, "--registry", REGISTRY, "--cache", freshCacheDir(), "--prefer-online"],
  ];
  if (resolved) attempts.push(["install", "-g", resolved.tarball]);

  for (const args of attempts) {
    const r = await run("npm", args, { capture: true });
    const tmpIdx = args.indexOf("--cache");
    if (tmpIdx >= 0) cleanupDir(args[tmpIdx + 1]);
    if (r.code === 0) return target ?? spec;
  }
  return null;
}

function freshCacheDir(): string {
  const dir = path.join(os.tmpdir(), `tokless-npm-${Date.now()}-${Math.random().toString(36).slice(2)}`);
  try {
    fs.mkdirSync(dir, { recursive: true });
  } catch {
  }
  return dir;
}

function cleanupDir(dir: string): void {
  try {
    fs.rmSync(dir, { recursive: true, force: true });
  } catch {
  }
}

File: internal/util/npminstall.go
package util

import (
	"encoding/json"
	"net/http"
	"net/url"
	"os"
	"time"
)

const registryURL = "https://registry.npmjs.org/"

type registryDoc struct {
	DistTags map[string]string `json:"dist-tags"`
	Versions map[string]struct {
		Version string `json:"version"`
		Dist    *struct {
			Tarball string `json:"tarball"`
		} `json:"dist"`
	} `json:"versions"`
}

// resolveFromRegistry resolves the version and tarball URL for a package spec.
func resolveFromRegistry(pkg, spec string) (string, string, bool) {
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(registryURL + url.QueryEscape(pkg))
	if err != nil {
		return "", "", false
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", "", false
	}
	var doc registryDoc
	if err := json.NewDecoder(resp.Body).Decode(&doc); err != nil {
		return "", "", false
	}
	version, ok := doc.DistTags[spec]
	if !ok {
		if _, exists := doc.Versions[spec]; exists {
			version = spec
		} else {
			return "", "", false
		}
	}
	vInfo, exists := doc.Versions[version]
	if !exists || vInfo.Dist == nil || vInfo.Dist.Tarball == "" {
		return "", "", false
	}
	return version, vInfo.Dist.Tarball, true
}

// NpmGlobalInstall installs an npm package globally.
func NpmGlobalInstall(pkg, spec string) (string, bool) {
	if spec == "" {
		spec = "latest"
	}
	var target string
	var tarball string
	resolvedVersion, tb, ok := resolveFromRegistry(pkg, spec)
	if ok {
		target = resolvedVersion
		tarball = tb
	} else {
		target = spec
	}

	atSpec := pkg + "@" + spec
	cacheDir := freshCacheDir()

	attempts := [][]string{
		{"install", "-g", atSpec},
		{"install", "-g", atSpec, "--prefer-online"},
		{"install", "-g", atSpec, "--registry", registryURL, "--cache", cacheDir, "--prefer-online"},
	}
	if ok && tarball != "" {
		attempts = append(attempts, []string{"install", "-g", tarball})
	}

	for _, args := range attempts {
		r := Run("npm", args, RunOptions{Capture: true})
		// Clean up the cache directory if this attempt used it.
		var hasCache bool
		for _, arg := range args {
			if arg == cacheDir {
				hasCache = true
				break
			}
		}
		if hasCache {
			cleanupDir(cacheDir)
		}
		if r.Code == 0 {
			return target, true
		}
	}
	return "", false
}

func freshCacheDir() string {
	dir, err := os.MkdirTemp("", "tokless-npm-*")
	if err != nil {
		return ""
	}
	return dir
}

func cleanupDir(dir string) {
	if dir != "" {
		_ = os.RemoveAll(dir)
	}
}
